tests/cupy_tests/test_cusolver.py

```diff
@@ -230,3 +230,50 @@ def test_gesv(self):
         x = cusolver.gesv(self.a, self.b)
         cupy.testing.assert_allclose(x, self.x_ref,
                                      rtol=self.tol, atol=self.tol)
+
+
+@testing.parameterize(*testing.product({
+    'dtype': [numpy.float32, numpy.float64, numpy.complex64, numpy.complex128],
+    'shape': [(4, 4), (5, 4), (4, 5)],
+    'nrhs': [None, 1, 4],
+}))
+@attr.gpu
+class TestGels(unittest.TestCase):
+    _tol = {'f': 1e-5, 'd': 1e-12}
+
+    def _make_array(self, shape):
+        a = testing.shaped_random(shape, cupy, dtype=self.r_dtype, scale=1)
+        return a
+
+    def _make_matrix(self, shape):
+        a = self._make_array(shape)
+        if self.dtype.char in 'FD':
+            a = a + 1j * self._make_array(shape)
+        return a
+
+    def setUp(self):
+        self.dtype = numpy.dtype(self.dtype)
+        if self.dtype.char in 'fF':
+            self.r_dtype = numpy.float32
+        else:
+            self.r_dtype = numpy.float64
+        m, n = self.shape
+        nrhs = 1 if self.nrhs is None else self.nrhs
+        a = self._make_matrix((m, n))
+        x = self._make_matrix((n, nrhs))
+        b = cupy.matmul(a, x)
+        b_shape = [m]
+        if self.nrhs is not None:
+            b_shape.append(nrhs)
+        self.a = a
+        self.b = b.reshape(b_shape)
+        if self.r_dtype == numpy.float32:
+            self.tol = self._tol['f']
+        elif self.r_dtype == numpy.float64:
+            self.tol = self._tol['d']
+
+    def test_gels(self):
+        x = cusolver.gels(self.a, self.b)
+        b = cupy.matmul(self.a, x)
+        cupy.testing.assert_allclose(b, self.b,
+                                     rtol=self.tol, atol=self.tol)
```

Comment:
The test seems too weak for under-determined system and too strong to pass for over-determined system.  How about comparing with `x, _, _, _ = numpy.linalg.lstsq(a, b, rcond=-1)`?

Reply:
If I understand correctly, the solution is not uniquely determined in the case of under-determined system. That said, the solution of `cusolver.gels()` and `xp.linalg.lstsq()` are not necessarily the same, so it is problematic to check the results by comparing `x`.

Reply:
Sorry that I missed how the matrices `a, b, x` are set up.  My suggestion is to generate `a, b` first.
- Resample `a` if `numpy.matrix_rank(a, eps) <= max(m, n)` to stabilize the tests or mark `@testing.condition.retry`.
- Comparing with `numpy.linalg.pinv(a) @ b` is also fine if your concern is about the specification of `numpy.linalg.lstsq`.

Reply:
I don't think there is anything wrong with the spec of `xp.linalg.lstsq`.
My point is that both `cusolver.gels` and `xp.linalg.lstsq` output the correct solution `x`, but their solutions will not match (if it is under-determined system). That is because `cusolver.gels` uses QR factorization based solver, while `xp.linalg.lstsq` uses SVD based solver.

Reply:
http://www.netlib.org/lapack/lug/node27.html says xGELS finds a minimum norm solution when m < n on the assumption that A has full rank.  Ignoring precision of float, the solution is unique and independent of solvers.

Reply:
Thanks for the information!
So, it is certainly strange if the solution of `gels` dose not match the solution of `lstsq`.
I think there may be a problem with my implementation, so let me check it out.

Reply:
Maybe `b = _cupy.zeros((n, nrhs), dtype=dtype, order='F')` fixes the issue.

Reply:
Thank you @toslunar . That was correct!